src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,437 @@
+                document, span, ExpressionBodyPreference.WhenPossible, cancellationToken).ConfigureAwait(false);
+
+            var toBlockBodyRefactorings = await ComputeRefactoringsAsync(
+                document, span, ExpressionBodyPreference.Never, cancellationToken).ConfigureAwait(false);
+
+            return toExpressionBodyRefactorings.AddRange(toBlockBodyRefactorings);
+        }
+
+        private async Task<ImmutableArray<CodeAction>> ComputeRefactoringsAsync(
+            Document document, TextSpan span, ExpressionBodyPreference option, CancellationToken cancellationToken)
+        {
+            if (span.Length > 0)
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            var position = span.Start;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var lambdaNode = root.FindToken(position).Parent.FirstAncestorOrSelf<LambdaExpressionSyntax>();
+            if (lambdaNode == null)
+            {
+                return ImmutableArray<CodeAction>.Empty;
+            }
+
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var result = ArrayBuilder<CodeAction>.GetInstance();
+            if (CanOfferUseExpressionBody(option, lambdaNode))
+            {
+                result.Add(new MyCodeAction(
+                    UseExpressionBodyTitle.ToString() + "-Ref",
+                    c => UpdateDocumentAsync(
+                        document, root, lambdaNode, c)));
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            if (CanOfferUseBlockBody(semanticModel, option, lambdaNode, cancellationToken))
+            {
+                result.Add(new MyCodeAction(
+                    UseBlockBodyTitle.ToString() + "-Ref",
+                    c => UpdateDocumentAsync(
+                        document, root, lambdaNode, c)));
+            }
+
+            return result.ToImmutableAndFree();
+        }
+
+        private async Task<Document> UpdateDocumentAsync(
+            Document document, SyntaxNode root, LambdaExpressionSyntax declaration, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var updatedDeclaration = Update(semanticModel, declaration);
+
+            var newRoot = root.ReplaceNode(declaration, updatedDeclaration);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        #endregion
+
+        #region shared feature code
+
+        private static ExpressionSyntax GetBodyAsExpression(LambdaExpressionSyntax declaration)
+            => declaration.Body as ExpressionSyntax;
+
+        private static bool CanOfferUseExpressionBody(
+            ExpressionBodyPreference preference, LambdaExpressionSyntax declaration)
```

Comment:
Because this is now just a single class, it is trivial to share functionality between the analyzer and refactoring.  Usually, the only solution here is to introduce another class with the common functionality needed by both of those.  This has led to the pattern of the 'Helper' class that goes along with most of code-style features.  This is no longer needed.